src/plugins/intel_cpu/src/nodes/matmul.cpp

```diff
@@ -456,24 +456,110 @@ std::pair<Shape, Shape> MatMul::makeDummyInputShapes(const Shape& in0, const Sha
     return {Shape(inDims0), Shape(inDims1)};
 }
 
+static bool SimplifyAs2D(dnnl::memory::desc& in0,
+                         dnnl::memory::desc& in1,
+                         dnnl::memory::desc& bias,
+                         dnnl::memory::desc& out) {
+    bool simplified = false;
+    auto in0_dims = in0.get_dims();
+    if (in0_dims.size() > 2 && bias.is_zero()) {
+        // in0: (bs0, bs1,..., m, k)
+        // in1: (bs0, bs1,..., k, n)
+        // bias:(bs0, bs1,..., m, n)
+        // out: (bs0, bs1,..., m, n)
+
+        // CVS-84056:
+        // on special cases, where:
+        //  in0: (bs0, bs1,..., m, k)
+        //  in1: (1,   1,  ..., k, n)
+        //  bias:(1,   1,  ..., 1, n)
+        //  out: (bs0, bs1,..., m, n)
+        // we can collapse batch dimensions into m using memory::desc::reshape()
+        // to form a equivalent but simpler 2D problem:
+        //  in0:  (bs0*bs1...*m, k)
+        //  in1:  (k, n)
+        //  bias: (1, n)
+        //  out:  (bs0*bs1...*m, n)
+        int rank = in0_dims.size();
+        auto in1_dims = in1.get_dims();
+        auto out_dims = out.get_dims();
+        do {
```

Comment:
It would be much better to refactor this part using the Guard Clauses. Just replace the `do while` and `break` with return statements:
```cpp
    if (in1_dims.size() != rank || out_dims.size() != rank) return simplified;
    ...
    for (int i = 0; i < rank-2; i++) {
    bs_collapsed *= in0_dims[i];
    if (in1_dims[i] != 1 || out_dims[i] != in0_dims[i])
        return simplified;
    }
    ...
    if (!new_in0) return simplified;
    ...
    /* and so on*/
```